File: Frontend/device_monitoring_frontend/src/services/alarmservice.ts
import {
  GetAlarmsQueryVariables,
  GetLatestAlarmForDeviceQueryVariables,
  getSdk,
  IgnoreAlarmMutationVariables,
  InvestigateAlarmMutationVariables,
  ResolveAlarmMutationVariables,
} from "@/graphql/generated/alarmService";
import { graphqlClient } from "./graphqlClient";

const sdk = getSdk(graphqlClient);

//Rest version
// export const getAlarmPanelData = async (alarmsFilters: any) => {
//   try {
//     const response = await axios.post(`${alarmServiceBaseURL}/api/Alarms/getAlarms`, alarmsFilters);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version
// const buildAlarmFilter = (filters: any) => {
//   return {
//     devices: Array.isArray(filters?.devices) ? filters.devices : [],
//     filterDateRange: Array.isArray(filters?.filterDateRange)
//       ? filters.filterDateRange
//       : [],
//   };
// };

// export const getAlarmPanelData = async (
//   alarmsFilters: any
// ) => {
//   const query = `
//     query ($filter: AlarmFilter!) {
//       alarms(filter: $filter) {
// id
// sourceDeviceMacId
// severity
// message
// raisedAt
// alarmState
// acknowledgedFrom
// isAcknowledged
// acknowledgedAt
// alarmComment
//       }
//     }
//   `;

//   const filter = buildAlarmFilter(alarmsFilters);

//   const data = await graphqlRequest<{ alarms: any[] }>(query, {
//     filter,
//   });

//   const response = {
//     data: data.alarms,
//   }
//   return response;
// };

//GraphQL version using generated sdk
export const getAlarmPanelData = async (
  filter: GetAlarmsQueryVariables["filter"]
) => {
  const data = await sdk.GetAlarms({ filter });
  const response = {
    data: data.alarms,
  };
  return response;
};

//Rest version

// export const getLatestAlarms = async () => {
//   try {
//     const response = await axios.get(`${alarmServiceBaseURL}/api/Alarms/getLatestAlarms`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version
// export const getLatestAlarms = async () => {
//   const query = `
//     query {
//       latestAlarms {
//         totalAlarms
//         alarms {
//           id
//           sourceDeviceMacId
//           severity
//           message
//           raisedAt
//           alarmState
//           acknowledgedFrom
//           isAcknowledged
//           acknowledgedAt
//           alarmComment
//         }
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ latestAlarms: any }>(query);

//   const response = {
//     data: data.latestAlarms,
//   };
//   return response;
// };

//GraphQL version using generated sdk
export const getLatestAlarms = async () => {
  const data = await sdk.GetLatestAlarms();
  const response = {
    data: data.latestAlarms,
  };
  return response;
};

//Rest version
// export const getLatestAlarmForDevice = async (deviceMacId : any) => {
//   try {
//     const response = await axios.get(`${alarmServiceBaseURL}/api/Alarms/getLatestAlarmForDevice/${deviceMacId}`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GRaphQL version
// export const getLatestAlarmForDevice = async (deviceMacId: any) => {
//   const query = `
//     query ($deviceMacId: String!) {
//       latestAlarmForDevice(deviceMacId: $deviceMacId) {
//         totalAlarms
//         alarm {
//           id
//           sourceDeviceMacId
//           severity
//           message
//           raisedAt
//           alarmState
//           acknowledgedFrom
//           isAcknowledged
//           acknowledgedAt
//           alarmComment
//         }
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ latestAlarmForDevice: any }>(query, {
//     deviceMacId,
//   });

//   const response = {
//     data: data.latestAlarmForDevice,
//   };
//   return response;
// };

export const getLatestAlarmForDevice = async (
  deviceMacId: GetLatestAlarmForDeviceQueryVariables["deviceMacId"]
) => {
  const data = await sdk.GetLatestAlarmForDevice({ deviceMacId });
  const response = {
    data: data.latestAlarmForDevice,
  };
  return response;
};

//Rest version
// export const investigateAlarm = async (alarmId: any) => {
//   try {
//     const response = await axios.put(`${alarmServiceBaseURL}/api/Alarms/investigateAlarm/${alarmId}`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version
// export const investigateAlarm = async (alarmId: any) => {
//   const mutation = `
//     mutation ($id: UUID!) {
//       investigateAlarm(id: $id) {
//         id
//         sourceDeviceMacId
//         alarmState
//         isAcknowledged
//         acknowledgedFrom
//         acknowledgedAt
//         severity
//         message
//         raisedAt
//         alarmComment
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ investigateAlarm: any }>(mutation, {
//     id: alarmId,
//   });

//   const response = {
//     data: data.investigateAlarm,
//   };
//   return response;
// };

//GraphQL version using generated sdk
export const investigateAlarm = async (
  id: InvestigateAlarmMutationVariables["id"]
) => {
  const data = await sdk.InvestigateAlarm({ id });
  const response = {
    data: data.investigateAlarm,
  };
  return response;
};

//Rest version
// export const resolveAlarm = async (alarmId: any, input : string) => {
//   const value = (input && input.length > 0) ? input :  "manual";
//   try {
//     const response = await axios.put(`${alarmServiceBaseURL}/api/Alarms/resolveAlarm/${alarmId}/${value}`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version
// export const resolveAlarm = async (alarmId: any, input: string) => {
//   const comment = input && input.length > 0 ? input : "manual";

//   const mutation = `
//     mutation ($id: UUID!, $comment: String!) {
//       resolveAlarm(id: $id, comment: $comment) {
//         id
//         sourceDeviceMacId
//         alarmState
//         isAcknowledged
//         acknowledgedFrom
//         acknowledgedAt
//         severity
//         message
//         raisedAt
//         alarmComment
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ resolveAlarm: any }>(mutation, {
//     id: alarmId,
//     comment,
//   });

//   const response = {
//     data: data.resolveAlarm,
//   };
//   return response;
// };

//GraphQL version using generated sdk
export const resolveAlarm = async (
  id: ResolveAlarmMutationVariables["id"],
  comment: ResolveAlarmMutationVariables["comment"]
) => {
  const data = await sdk.ResolveAlarm({ id, comment });
  const response = {
    data: data.resolveAlarm,
  };
  return response;
};

//Rest version
// export const ignoreAlarm = async (alarmId: any, input : string) => {
//   const value = (input && input.length > 0) ? input :  "manual";
//   try {
//     const response = await axios.delete(`${alarmServiceBaseURL}/api/Alarms/ignoreAlarm/${alarmId}/${value}`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version
// export const ignoreAlarm = async (alarmId: any, input: string) => {
//   const comment = input && input.length > 0 ? input : "manual";

//   const mutation = `
//     mutation ($id: UUID!, $comment: String!) {
//       ignoreAlarm(id: $id, comment: $comment) {
//         id
//         sourceDeviceMacId
//         alarmState
//         isAcknowledged
//         acknowledgedFrom
//         acknowledgedAt
//         severity
//         message
//         raisedAt
//         alarmComment
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ ignoreAlarm: any }>(mutation, {
//     id: alarmId,
//     comment,
//   });

//   const response = {
//     data: data.ignoreAlarm,
//   };
//   return response;
// };

//GraphQL version using generated sdk
export const ignoreAlarm = async (
  id: IgnoreAlarmMutationVariables["id"],
  comment: IgnoreAlarmMutationVariables["comment"]
) => {
  const data = await sdk.IgnoreAlarm({ id, comment });
  const response = {
    data: data.ignoreAlarm,
  };
  return response;
};

//Rest version alarm states
// export const getAlarmStates = async () => {
//   try {
//     const response = await axios.get(`${alarmServiceBaseURL}/api/Alarms/getAlarmStates`);
//     return response;
//   } catch (error : any) {
//      handleAxiosError(error);
//   }
// };

//GraphQL version alarm states
// export const getAlarmStates = async () => {
//   const query = `
//     query {
//       alarmStates {
//         id
//         name
//       }
//     }
//   `;

//   const data = await graphqlRequest<{ alarmStates: any[] }>(query);

//   const response = {
//     data: data.alarmStates,
//   };
//   return response;
// };

export const getAlarmStates = async () => {
  const data = await sdk.GetAlarmStates();
  const response = {
    data: data.alarmStates,
  };
  return response;
};

File: Backend/AlarmService/Infrastructure/Services/AlarmService.cs
using System.Globalization;
using System.Security.Claims;
using System.Text.Encodings.Web;
using System.Text.Json;
using Application.Dtos;
using Common.Helper_Classes;
using Domain.Entities;
using Domain.Interface;
using Infrastructure.Persistence;
using Infrastructure.RealTime;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;
using static Microsoft.EntityFrameworkCore.DbLoggerCategory;

namespace Infrastructure.Services
{
    public class AlarmService : IAlarmService
    {
        private readonly AlarmDbContext _context;
        private readonly IHubContext<AlertHub> _hubContext;

        public AlarmService(AlarmDbContext context, IHubContext<AlertHub> hubContext)
        {
            _context = context;
            _hubContext = hubContext;
        }

        public async Task<IEnumerable<GetAlarmDto>> GetAlarms(AlarmFilter filter)
        {
            var ignoredStateId = await _context.AlarmStates.Where(state => state.Name == "Ignored").Select(state => state.Id).FirstOrDefaultAsync();
            var alarms = _context.Alarms.Where(alarm => alarm.StateId != ignoredStateId).Include(a => a.State).AsQueryable();
            if (filter.Devices is not null && filter.Devices.Count > 0)
                alarms = alarms.Where(a => filter.Devices.Contains(a.SourceDeviceMacId));

            if (filter.FilterDateRange?.Length == 2 &&
                DateTime.TryParse
                        (filter.FilterDateRange[0],
                        CultureInfo.InvariantCulture,
                        DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                        out var startDate) &&
                DateTime.TryParse
                        (filter.FilterDateRange[1],
                        CultureInfo.InvariantCulture,
                        DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                        out var endDate))
            {
                alarms = alarms.Where(a => a.RaisedAt >= startDate && a.RaisedAt <= endDate);
            }

            var formattedAlarms = await alarms.Select(alarm => new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId,
                AlarmState = alarm.State.Name,
                AlarmComment = alarm.Comment,
                AcknowledgedFrom = alarm.AcknowledgedFrom
            }).ToListAsync();

            return formattedAlarms;
        }

        public async Task<GetAlarmDto> GetAlarm(Guid id)
        {
            var alarm = await _context.Alarms.FindAsync(id);

            if (alarm == null)
                throw new CustomException(404, "Alarm not found");

            return new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId
            };
        }

        public async Task<string> PutAlarm(Guid id, Alarm alarm)
        {
            if (id != alarm.Id)
                throw new CustomException(400, "Alarm id's don't match");

            _context.Entry(alarm).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException)
            {
                if (!AlarmExists(id))
                {
                    throw new CustomException(404, "Alarm not found");
                }
                else
                {
                    throw;
                }
            }

            return "Alarm updated successfully";
        }

        public async Task<Alarm> PostAlarm(PostAlarmDto alarmDto)
        {
            var alarm = new Alarm
            {
                SourceDeviceMacId = alarmDto.SourceDeviceMacId,
                Message = alarmDto.Message,
                Severity = alarmDto.Severity
            };

            _context.Alarms.Add(alarm);
            await _context.SaveChangesAsync();

            return alarm;
        }

        public async Task<GetAlarmDto> IgnoreAlarm(Guid id, string comment)
        {
            var alarm = await _context.Alarms.SingleOrDefaultAsync(alarm => alarm.Id == id);
            if (alarm is null)
                throw new CustomException(400, "Alarm not found");

            if (alarm.IsAcknowledged is false)
            {
                alarm.IsAcknowledged = true;
                alarm.AcknowledgedAt = DateTime.Now;
                alarm.AcknowledgedFrom = "Ignored without investigation";
            }
            else
            {
                alarm.AcknowledgedFrom = "Ignored with investigation";
            }

            var ignoredState = await _context.AlarmStates.SingleOrDefaultAsync(alarm => alarm.Name == "Ignored");
            alarm.StateId = ignoredState?.Id??1002; //1002 is hardcoded here later will throw exception or handle it here
            alarm.Comment = (comment != "manual" && comment.Length > 0) ? comment : "Manually marked as Ignored";

            _context.Entry(alarm).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException ex)
            {
                throw new CustomException(500, ex.Message);
            }

            var mainPageUpdates = await GetLatestFiveAlarms();
            var serializedData = JsonSerializer.Serialize(mainPageUpdates, new JsonSerializerOptions
            {
                Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
                WriteIndented = true,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            });
            await _hubContext.Clients.All.SendAsync("ReceiveMainPageUpdates", serializedData);

            var propertyPanelAlarm = await GetLatestAlarmForDevice(alarm.SourceDeviceMacId);
            await _hubContext.Clients.Group($"Alarm-{alarm.SourceDeviceMacId}").SendAsync("ReceivePropertyPanelAlarmUpdates", JsonSerializer.Serialize(propertyPanelAlarm, new JsonSerializerOptions
            {
                Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = true
            }));

            return new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId,
                AlarmState = alarm.State.Name,
                AlarmComment = alarm.Comment,
                AcknowledgedFrom = alarm.AcknowledgedFrom
            };
        }

        private bool AlarmExists(Guid id)
        {
            return _context.Alarms.Any(e => e.Id == id);
        }

        public async Task<IEnumerable<GetAlarmDto>> GetAlarmsByDeviceId(string id)
        {
            var ignoredStateId = await _context.AlarmStates.Where(state => state.Name == "Ignored").Select(state => state.Id).FirstOrDefaultAsync();
            var alarms = await _context.Alarms.Where(alarm => alarm.SourceDeviceMacId == id && alarm.StateId != ignoredStateId).Include(a => a.State).ToListAsync();

            var formattedAlarms = new List<GetAlarmDto>();
            alarms.ForEach(alarm => {
                formattedAlarms.Add(new GetAlarmDto
                {
                    Id = alarm.Id,
                    AcknowledgedAt = alarm.AcknowledgedAt,
                    IsAcknowledged = alarm.IsAcknowledged,
                    Message = alarm.Message,
                    RaisedAt = alarm.RaisedAt,
                    Severity = alarm.Severity.ToString(),
                    SourceDeviceMacId = alarm.SourceDeviceMacId,
                    AlarmState = alarm.State.Name,
                    AlarmComment = alarm.Comment,
                    AcknowledgedFrom = alarm.AcknowledgedFrom
                });
            });

            return formattedAlarms;
        }

        public async Task<GetAlarmDto> InvestigateAlarm(Guid alarmId)
        {
            var alarm = await _context.Alarms.SingleOrDefaultAsync(alarm => alarm.Id == alarmId);
            if (alarm is null)
                throw new CustomException(400, "Alarm not found");

            alarm.IsAcknowledged = true;
            alarm.AcknowledgedAt = DateTime.Now;

            var investigateState = await _context.AlarmStates.SingleOrDefaultAsync(alarm => alarm.Name == "Investigating");
            alarm.StateId = investigateState?.Id ?? 2; //2 is hardcoded here later will throw exception or handle it here
            alarm.Comment = "Under Investigation";
            alarm.AcknowledgedFrom = "Alarm acknowledged";

            _context.Entry(alarm).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException ex)
            {
                throw new CustomException(500, ex.Message);
            }

            return new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId,
                AlarmState = alarm.State.Name,
                AlarmComment = alarm.Comment,
                AcknowledgedFrom = alarm.AcknowledgedFrom
            };
        }

        public async Task<GetLatestAlarmsDto> GetLatestFiveAlarms()
        {
            var ignoredStateId = await _context.AlarmStates.Where(state => state.Name == "Ignored").Select(state => state.Id).FirstOrDefaultAsync();
            var alarms = await _context.Alarms.Where(alarm => alarm.StateId != ignoredStateId).Include(a => a.State).OrderByDescending(a => a.RaisedAt).Take(5).ToListAsync();
            var totalAlarms = await _context.Alarms.Where(alarm => alarm.StateId != ignoredStateId).CountAsync();

            var formattedAlarms = new List<GetAlarmDto>();
            alarms.ForEach(alarm => {
                formattedAlarms.Add(new GetAlarmDto
                {
                    Id = alarm.Id,
                    AcknowledgedAt = alarm.AcknowledgedAt,
                    IsAcknowledged = alarm.IsAcknowledged,
                    Message = alarm.Message,
                    RaisedAt = alarm.RaisedAt,
                    Severity = alarm.Severity.ToString(),
                    SourceDeviceMacId = alarm.SourceDeviceMacId,
                    AlarmState = alarm.State.Name,
                    AlarmComment = alarm.Comment
                });
            });

            return new GetLatestAlarmsDto
            {
                TotalAlarms = totalAlarms,
                Alarms = formattedAlarms
            };
        }

        public async Task<GetLatestAlarmForDeviceDto> GetLatestAlarmForDevice(string deviceMacId)
        {
            var ignoredStateId = await _context.AlarmStates.Where(state => state.Name == "Ignored").Select(state => state.Id).FirstOrDefaultAsync();
            var alarms = _context.Alarms.Where(a => a.SourceDeviceMacId == deviceMacId && a.StateId != ignoredStateId)?.Include(a => a.State).OrderByDescending(a => a.RaisedAt).AsQueryable();
            var totalAlarms = await alarms.CountAsync();
            var alarm = await alarms.FirstOrDefaultAsync();
            if (alarm == null)
                throw new CustomException(404, "Alarm for this device not found");

            var formattedAlarm = new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId,
                AlarmState = alarm.State.Name
            };

            return new GetLatestAlarmForDeviceDto
            {
                TotalAlarms = totalAlarms,
                Alarm = formattedAlarm
            };
        }

        public async Task<IEnumerable<GetAlarmStatesDto>> GetAlarmStates()
        {
            var states = await _context.AlarmStates.Select(state => new GetAlarmStatesDto
            {
                Id= state.Id,
                Name = state.Name,
            }).ToListAsync();

            return states;
        }

        public async Task<GetAlarmDto> ResolveAlarm(Guid alarmId, string comment)
        {
            var alarm = await _context.Alarms.SingleOrDefaultAsync(alarm => alarm.Id == alarmId);
            if (alarm is null)
                throw new CustomException(400, "Alarm not found");

            if(alarm.IsAcknowledged is false)
            {
                alarm.IsAcknowledged = true;
                alarm.AcknowledgedAt = DateTime.Now;
                alarm.AcknowledgedFrom = "Resolved without investigation";
            }
            else
            {
                alarm.AcknowledgedFrom = "Resolved with investigation";
            }

                var resolvedState = await _context.AlarmStates.SingleOrDefaultAsync(alarm => alarm.Name == "Resolved");
            alarm.StateId = resolvedState?.Id ?? 3; //3 is hardcoded here later will throw exception or handle it here
            alarm.Comment = (comment != "manual" && comment.Length > 0) ? comment : "Manually marked as Resoved";
            alarm.ResolvedAt = DateTime.Now;
            

            _context.Entry(alarm).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException ex)
            {
                throw new CustomException(500, ex.Message);
            }

            return new GetAlarmDto
            {
                Id = alarm.Id,
                AcknowledgedAt = alarm.AcknowledgedAt,
                IsAcknowledged = alarm.IsAcknowledged,
                Message = alarm.Message,
                RaisedAt = alarm.RaisedAt,
                Severity = alarm.Severity.ToString(),
                SourceDeviceMacId = alarm.SourceDeviceMacId,
                AlarmState = alarm.State.Name,
                AlarmComment = alarm.Comment,
                AcknowledgedFrom = alarm.AcknowledgedFrom
            };
        }
    }
}
